lib/storage/bucket.js

```diff
@@ -723,68 +734,255 @@ Bucket.prototype.setMetadata = function(metadata, callback) {
  *   // Your bucket now contains:
- *   // - "existing-file.png" (with the contents of `local-img.png')
+ *   //   - "existing-file.png" (with the contents of `local-img.png')
  *
  *   // Note:
- *   // The `newFile` parameter is equal to `file`.
+ *   // The `newFile` parameter is equal to `options.destination`.
+ * });
+ *
+ * //-
+ * // For the power users, glob patterns are also supported, using
+ * // <a href="https://github.com/sindresorhus/globby">sindresorhus/globby</a>
+ * // under the hood.
+ * //
+ * // NOTE: All of the options globby accepts can be provided with
+ * // `options.globOptions`. For a full list of options, see
+ * // <a href="http://goo.gl/14UhaI">node-glob's options object</a>.
+ * //-
+ * bucket.upload('/Users/stephen/Desktop/*.{jpg|png}', function(errors, files) {
+ *   // `errors` will always be an array.
+ *   // `files` is an array of all successfully uploaded files that the glob
+ *   // pattern matched.
+ * });
+ *
+ * //-
+ * // If you're uploading many files, you may wish to suppress errors until all
+ * // of the files have been processed.
+ * //-
+ * var options = {
+ *   force: true
+ * };
+ *
+ * bucket.upload('/Users/stephen/Photos/*', options, function(errors, files) {
+ *   // `errors` will always be an array.
+ *   // `files` is an array of all successfully uploaded files that the glob
+ *   // pattern matched.
  * });
  */
-Bucket.prototype.upload = function(localPath, options, callback) {
+Bucket.prototype.upload = function(pattern, options, callback) {
+  var self = this;
+
   if (util.is(options, 'function')) {
     callback = options;
     options = {};
   }
 
-  var newFile;
-  if (options.destination instanceof File) {
-    newFile = options.destination;
-  } else if (util.is(options.destination, 'string')) {
-    // Use the string as the name of the file.
-    newFile = this.file(options.destination);
-  } else {
-    // Resort to using the name of the incoming file.
-    newFile = this.file(path.basename(localPath));
+  options = options || {};
+
+  var errors = [];
+  var files = [];
+
+  var globOptions = extend({}, options.globOptions, { nodir: true });
+
+  globby(pattern, globOptions, function(err, filePaths) {
+    if (err) {
+      callback(err);
+      return;
+    }
+
+    var uploadFileFns = filePaths.map(function(filePath) {
+      return function(next) {
+        var fileName;
+
+        if (options.basePath) {
+          fileName = path.relative(options.basePath, filePath);
+        } else {
+          fileName = path.basename(filePath);
+        }
+
+        var opts = extend({}, options, { destination: fileName });
+
+        self.uploadFile(filePath, opts, function(err, file) {
+          if (err) {
+            errors.push(err);
+          } else {
+            files.push(file);
+          }
+
+          next(options.force ? null : err);
+        });
+      };
+    });
+
+    async.parallelLimit(uploadFileFns, MAX_PARALLEL_UPLOADS, function() {
```

Comment:
Can you just use `async.eachLimit(filePaths, MAX_PARALLEL_UPLOADS, function(next) {}, function(){});` (where `function(next) {}` is the function returned by the map call, that way for N files, we don't have to create an array of N functions which we then iterate through, we just create one function and reuse it for each of the filenames?